src/uptane.rs

```diff
@@ -33,43 +30,90 @@ impl Display for Service {
+            repo_server:     config.uptane.repo_server.clone(),
+            metadata_path:   config.uptane.metadata_path.clone(),
+
+            primary_ecu: config.uptane.primary_ecu_serial.clone(),
+            private_key: PrivateKey {
+                keyid: digest::digest(&digest::SHA256, &private_key)
+                    .as_ref()
+                    .iter()
+                    .map(|b| format!("{:02x}", b))
+                    .collect::<String>(),
+                der_key: private_key
+            },
+            sig_type: SignatureType::RsaSsaPss,
+
+            director_verifier: Verifier::default(),
+            repo_verifier:     Verifier::default(),
+        };
+
+        uptane.add_root_keys(Service::Director)?;
+        uptane.add_root_keys(Service::Repo)?;
+        uptane.update_root_keys(client, Service::Director)?;
+        uptane.update_root_keys(client, Service::Repo)?;
+
+        Ok(uptane)
     }
 
     /// Returns a URL based on the uptane service.
     fn endpoint(&self, service: Service, endpoint: &str) -> Url {
-        let cfg = &self.config;
         match service {
-            Service::Director => cfg.director_server.join(&format!("/{}", endpoint)),
-            Service::Repo     => cfg.repo_server.join(&format!("/{}/{}", self.device_id, endpoint))
+            Service::Director => self.director_server.join(&format!("/{}", endpoint)),
+            Service::Repo     => self.repo_server.join(&format!("/{}", endpoint))
+        }
+    }
+
+    /// Returns the respective key verifier for an uptane service.
+    fn verifier(&mut self, service: Service) -> &mut Verifier {
+        match service {
+            Service::Director => &mut self.director_verifier,
+            Service::Repo     => &mut self.repo_verifier
         }
     }
 
+    /// Add the keys from a service's local `root.json` metadata to its verifier.
+    fn add_root_keys(&mut self, service: Service) -> Result<(), Error> {
+        let json = Util::read_file(&format!("{}/{}/root.json", self.metadata_path, service))?;
+        let signed = json::from_slice::<TufSigned>(&json)?;
+        let data = json::from_value::<RoleData>(signed.signed)?;
+        for (role, meta) in data.roles.ok_or(Error::UptaneMissingRoles)? {
+            self.verifier(service).add_meta(role, meta)?;
+        }
+        for (id, key) in data.keys.ok_or(Error::UptaneMissingKeys)? {
+            self.verifier(service).add_key(id, key)?;
+        }
+        Ok(())
+    }
+
+    /// Download, validate and replace the latest `root.json` metadata.
+    fn update_root_keys(&mut self, client: &Client, service: Service) -> Result<(), Error> {
+        let json = self.get(client, service, "root.json")?;
+        let signed = json::from_slice::<TufSigned>(&json)?;
+        let verified = self.verify_metadata(service, RoleName::Root, signed)?;
+        if verified.is_new() {
+            Util::write_file(&format!("{}/{}/root.json", self.metadata_path, service), &json)?;
```

Comment:
The spec suggests keeping old metadata around, especially the roots. You should save them as N.root.json so that full trust can be rebuilt from the original if need be.